Fixed #17419 -- Added json_script template filter.

jonashaag opened the pull request:
https://code.djangoproject.com/ticket/17419

orf commented on django/template/defaultfilters.py:
```diff
@@ -82,6 +82,12 @@ def escapejs_filter(value):
 
 @register.filter(is_safe=True)
+def json_tag(value, element_id):
+    """Output value JSON-encoded, wrapped in a <script type="application/javascript"> HTML tag."""
```
Should be `application/json`

jonashaag commented:
Why has GitHub deleted the circular import question? Anyways, yes:

```
>>> import django.utils.html
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/j/dev/django/django/utils/html.py", line 15, in <module>
    from django.core.serializers.json import DjangoJSONEncoder
  File "/Users/j/dev/django/django/core/serializers/__init__.py", line 23, in <module>
    from django.core.serializers.base import SerializerDoesNotExist
  File "/Users/j/dev/django/django/core/serializers/base.py", line 6, in <module>
    from django.db import models
  File "/Users/j/dev/django/django/db/models/__init__.py", line 3, in <module>
    from django.db.models.aggregates import *  # NOQA
  File "/Users/j/dev/django/django/db/models/aggregates.py", line 5, in <module>
    from django.db.models.expressions import Case, Func, Star, When
  File "/Users/j/dev/django/django/db/models/expressions.py", line 7, in <module>
    from django.db.models import fields
  File "/Users/j/dev/django/django/db/models/fields/__init__.py", line 11, in <module>
    from django import forms
  File "/Users/j/dev/django/django/forms/__init__.py", line 6, in <module>
    from django.forms.boundfield import *  # NOQA
  File "/Users/j/dev/django/django/forms/boundfield.py", line 3, in <module>
    from django.forms.utils import flatatt, pretty_name
  File "/Users/j/dev/django/django/forms/utils.py", line 7, in <module>
    from django.utils.html import escape, format_html, format_html_join, html_safe
ImportError: cannot import name 'escape'
```

claudep commented on docs/ref/templates/builtins.txt:
```diff
@@ -1628,6 +1628,34 @@ For example::
+
+If ``value`` is a the dictionary ``{'hello': ['world', 'wide']}``, the output will be::
+
+    <script id="value_json" type="application/json">{"hello": ["world", "wide"]}</script>
```
Shouldn't the documentation talk about escaped characters, either by including one in the example snippet or by a separate note below?

ngnpope commented on django/utils/html.py:
```diff
@@ -72,6 +73,28 @@ def escapejs(value):
+_json_tag_escapes = {
+    ord('>'): '\\u003E',
+    ord('<'): '\\u003C',
+    ord('&'): '\\u0026'
```
Please add a trailing comma.

ngnpope commented on docs/ref/templates/builtins.txt:
```diff
@@ -1628,6 +1628,34 @@ For example::
+
+For example::
+
+    {{ value|json_tag:"value_json" }}
```
Perhaps a better id than `value_json` would make the example clearer, e.g. `hello-data`?

claudep commented:
Sorry, some test failures will be fixed by #9243.

jonashaag commented:
Unfortunately the `docs` CI task doesn't say which words are misspelled.

pope1ni commented:
If you build the documentation locally, which ought to be done before pushing, then sphinx will output misspelled / unknown words. See the [documentation](https://docs.djangoproject.com/en/2.0/internals/contributing/writing-documentation/#getting-started-with-sphinx).

jonashaag commented:
\o/

Squashed + rebased to latest master.

MarkusH commented:
@apollo13 Following our discussion when prepping the Django security talk, I think this is the general idea we had there.

jonashaag commented:
The test failure seems to be unrelated to this PR.

apollo13 commented:
How where the characters to escape chosen? I am asking because to the best of my knowledge, a script tag switches the parsing modes of browsers and they only end that block when `</script>` (+/- whitespace) is encountered. In that sense replacing `<` should be enough.

jonashaag commented:
> How where the characters to escape chosen? I am asking because to the best of my knowledge, a script tag switches the parsing modes of browsers and they only end that block when </script> (+/- whitespace) is encountered. In that sense replacing < should be enough.

I just took them from the original patch, see https://code.djangoproject.com/ticket/17419#comment:27. It seems the characters are taken from Ruby on Rails `json_escape`: http://api.rubyonrails.org/classes/ERB/Util.html#method-c-json_escape

It might very well be that escaping `<` suffices but it also doesn't hurt to escape any non-JSON characters (i.e. everything except `{}":,\` etc)

jonashaag commented:
Rebased, tests now pass

jonashaag commented:
@apollo13 Do you have any open questions that I should address?

apollo13 commented on django/utils/html.py:
```diff
@@ -77,6 +78,28 @@ def escapejs(value):
+    This will escape all the HTML/XML special characters with their unicode
+    escapes, so it is safe to be output anywhere except for inside a tag
+    attribute. It will wrap the escaped json in a script tag.
+    See https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet, rule #3.1 for details.
```
I'd drop that comment since it doesn't really escape according to the rules suggested there.

apollo13 commented on django/utils/html.py:
```diff
@@ -77,6 +78,28 @@ def escapejs(value):
     return mark_safe(str(value).translate(_js_escapes))
 
 
+_json_tag_escapes = {
```
I'd still like to see an argument here why we replace those, the reference to http://api.rubyonrails.org/classes/ERB/Util.html#method-c-json_escape does not really count because they insert it in a normal script tag which probably has different parsing rules again. Following the argument in https://github.com/django/django/pull/9235#issuecomment-337607943 I'd also escape single quotes then. Do we have any easily readable document on how browsers are supposed to parse that :/

apollo13 commented:
The reason for my reservations with the current escaping rules is that I'd like be able to explain them easily, which is currently not the case…

https://mathiasbynens.be/notes/etago has some interesting details about parsing modes and what not.

I think we should drop escaping of `&` unless we can find any issue with it. The same probably applies to `>`; but it is part of ending an HTML comment which might cause some problems (comments in script are still somewhat handled…), so I'd leave it in there.

jonashaag commented:
> I think we should drop escaping of & unless we can find any issue with it.

Does it hurt though? I'd rather escape all the things in case we forget to think about some special case.

apollo13 commented:
> Does it hurt though? I'd rather escape all the things in case we forget to think about some special case.

Then why are we not escaping stuff like single quotes? The thing is that you are already having a hard time to argue __why__ you are choosing those characters to escape, making up arbitrary rules doesn't make it any better.

jonashaag commented:
> Then why are we not escaping stuff like single quotes? The thing is that you are already having a hard time to argue why you are choosing those characters to escape, making up arbitrary rules doesn't make it any better.

Sure, we can escape single quotes as well, no problem.

From what I understand it's enough to escape `<` because this accounts for `</script>` and `<!--`. I think it can't hurt to add some other things that look like they might be harmful to the list as well, which includes (gut feeling) `&>'\`. If you prefer to not have them in that's fine as well.

The list in the patch is taken from Ruby on Rails; these guys seem to have been using that for years. From my experiments it shows that `<script type=X>` is parsed the same as `<script type=Y>`. This is also confirmed in the HTML standard, see https://html.spec.whatwg.org/multipage/syntax.html#cdata-rcdata-restrictions.

I have done some testing with auto-generated possibly-harmful strings:

https://gist.github.com/jonashaag/627bde9ee79be0851c4e1a98756520e5

This thing generates random evil-ish strings, stuffs them in `<script>` tags and compares their HTML-parsed result with the original string. If the strings are identical, then we've correctly escaped everything. My results are that it suffices to escape `<`. I also compared `<script type=application/json>` to `<script type=text/javascript>` which are indeed parsed identically. I also compared HTML4 to HTML5 doctype behaviour. As noted in the link provided in your previous comment these show identical behaviour because browsers never implemented the "correct" HTML4 rules.

apollo13 commented:
@timgraham any thoughts on the documentation?

Count that as approval from me.

timgraham commented on tests/template_tests/filter_tests/test_js_json.py:
```diff
@@ -0,0 +1,52 @@
+    def test_json_tag(self):
+        output = self.engine.render_to_string(
+            'json-tag01',
+            {'value': {'a': 'testing\r\njson \'string" <b>escaping</b>'}})
```
Put the closing ) on the next line.

timgraham commented on tests/template_tests/filter_tests/test_js_json.py:
```diff
@@ -0,0 +1,52 @@
+            'json-tag01',
+            {'value': {'a': 'testing\r\njson \'string" <b>escaping</b>'}})
+        self.assertEqual(output, '<script id="test_id" type="application/json">'
+                                 '{"a": "testing\\r\\njson \'string\\" '
```
Use hanging indent as the other tests do.

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -1628,6 +1628,38 @@ For example::
+    var el = document.getElementById('hello-data');
+    var value = JSON.parse(el.textContent || el.innerText);
+
+Possible XSS attacks are mitigated by escaping the characters "<", ">" and "&". For example if ``value`` is ``{'hello': 'world</script>&amp;'}``, the filter's output is::
```
Wrap docs at 79 characters please.

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -1628,6 +1628,38 @@ For example::
+
+This is compatible with a strict Content Security Policy that prohibits all
+in-page script execution and maintains a clean separation between passive data
+and executable code.  See https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet#RULE_.233.1_-_HTML_escape_JSON_values_in_an_HTML_context_and_read_the_data_with_JSON.parse for more details.
```
Please don't put raw URLs in text. That being said, there's a comment from Florian in the function's docstring that questions if this URL should be referenced at all.

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -1628,6 +1628,38 @@ For example::
 If ``value`` is ``"testing\r\njavascript \'string" <b>escaping</b>"``,
 the output will be ``"testing\\u000D\\u000Ajavascript \\u0027string\\u0022 \\u003Cb\\u003Eescaping\\u003C/b\\u003E"``.
 
+.. templatefilter:: json_tag
```
Please alphabetize this after the "join" tag.

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -1628,6 +1628,38 @@ For example::
+
+.. versionadded:: 2.1
+
+Outputs an object JSON-encoded, wrapped in a <script> tag ready for use with JavaScript.
```
<script> should have backticks around it.

timgraham commented on docs/ref/templates/builtins.txt:
```diff
@@ -1628,6 +1628,38 @@ For example::
+
+    {{ value|json_tag:"hello-data" }}
+
+If ``value`` is a the dictionary ``{'hello': 'world'}``, the output will be::
```
Use a single : and `.. code-block:: javascript` for proper highlighting.

timgraham commented on tests/template_tests/filter_tests/test_js_json.py:
```diff
@@ -0,0 +1,52 @@
+from ..utils import setup
+
+
+class JsJsonTests(SimpleTestCase):
```
Where does the name "test_js_json.py". I'd follow the naming of the tag "test_json_tag.py", although I wonder what "tag" in "json_tag" refers to?

timgraham commented on django/utils/html.py:
```diff
@@ -77,6 +78,28 @@ def escapejs(value):
+
+def json_tag(value, element_id):
+    """
+    This will escape all the HTML/XML special characters with their unicode
```
drop 'this will" / "it will"

jonashaag commented:
- Moved function tests to `test_html.py`
- Added release note
- Docs fixes
- Renamed to `json_script`. The thing was originally named `js_json`, and someone else suggested `json_tag`. I simply used that discussion as the basis for this PR. I have no affiliation with the same whatsoever.
- Changed a few code indents. I left a few code lines longer than 79 chars because splitting them would hurt readability.

jonashaag commented:
The explicit tests for lazy strings were in the original patch. I'm not sure why they exist; I don't see why lazy strings should behave differently here.
